### PR TITLE
Add QAD example for Alpamayo

### PR DESCRIPTION
**Pull request by rohansjoshi**
### What does this PR do?

Add `examples/alpamayo/qad.py` which distills the quantized Alpamayo checkpoint's VLM against the FP16 VLM of the original with ModelOpt's QADTrainer, and shards student and teacher with FSDP2 for multi-GPU runs. Only the VLM is trained; the action expert stays frozen.

Type of change: new example

### Usage

```
torchrun --standalone --nproc_per_node 8 qad.py \
    --student_ckpt ./alpamayo-auto \
    --output_dir ./alpamayo-auto-qad \
    --parquet ./train_clips.parquet \
    --max_steps 500 --fsdp2 --grad_ckpt --export
```

### Testing
Tested end-to-end on public Alpamayo-1 checkpoint

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?: ✅ / ❌ / N/A 
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: ✅ / ❌ / N/A 
- Did you write any new necessary tests?: ✅ / ❌ / N/A 
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ✅ / ❌ / N/A 
- Did you get Claude approval on this PR?: ✅ / ❌ / N/A 

## Summary by CodeRabbit

- **New Features**
  - Added a quantization-aware distillation workflow for AlpamayoR1 models.
  - Supports prompt-only or rollout-inclusive distillation, FSDP2 training, dataset slicing, revision pinning, and checkpoint resumption.
  - Supports exporting trained models as complete, reloadable AlpamayoR1 checkpoints.
  - Added optional vision-parameter freezing, trajectory-history fusion, gradient checkpointing, evaluation, and synchronized training cadence.

- **Documentation**
  - Expanded the Alpamayo guide with setup, training, dataset, and export instructions.
  - Clarified sensitivity-based quantization behavior.
  - Added a version 0.47 quantization changelog entry.

### CODE REVIEW

File: examples/alpamayo/qad.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+        help="pin the PhysicalAI-AV dataset revision; required to read a cache offline",
+    )
+    ap.add_argument("--no_fuse_traj", dest="fuse_traj", action="store_false", default=True)
+
+    ap.add_argument("--max_steps", type=int, default=100)
+    ap.add_argument("--lr", type=float, default=1e-6)
+    ap.add_argument("--warmup_ratio", type=float, default=0.05)
+    ap.add_argument("--grad_accum", type=int, default=1)
+    ap.add_argument("--temperature", type=float, default=1.0)
+    ap.add_argument("--logging_steps", type=int, default=1)
```
Comment (cjluo-nv) [inline examples/alpamayo/qad.py:240]: > _Bot comment._

The defaults make validation overlap training: the README command uses `--limit_train 2000`, while the default validation slice is indices 12–15. Thus enabling `--eval_steps` on the documented command reports KD/CE losses on training examples. Please either choose a non-overlapping validation default based on the end of the train slice, require explicit validation offsets, or reject overlapping slices.

File: examples/alpamayo/qad.py [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+    full = AlpamayoR1.from_pretrained(student_ckpt, dtype=dtype)
+
+    shards = sorted(glob.glob(os.path.join(trained_vlm_dir, "model*.safetensors")))
+    if not shards:
+        raise FileNotFoundError(f"no model*.safetensors found in {trained_vlm_dir}")
+    state_dict = {}
+    for shard in shards:
+        state_dict.update(load_file(shard))
+
+    missing, unexpected = full.vlm.load_state_dict(state_dict, strict=False)
```
Comment (cjluo-nv) [inline examples/alpamayo/qad.py:183]: > _Bot comment._

Please don't merely log an incompatible trained state dict and continue saving. A prefix/layout mismatch could leave most or all of `full.vlm` at the original checkpoint weights while still producing a reloadable “export.” At minimum, fail on unexpected keys and on missing trainable/model parameter keys (while explicitly allowlisting any intentionally absent non-persistent quantizer buffers). This critical export validation should also have a small unit test using sharded safetensors, including a mismatched-key failure case.

**Review by cjluo-nv — COMMENTED** [review]
> _Bot review (gpt-5.6-sol) — DM the bot to share feedback._

The example is cohesive and reuses the existing Alpamayo input helpers and QADTrainer; the new file's NVIDIA license header matches the repository's canonical LICENSE_HEADER. However, the documented/default validation split overlaps the training split, and export can silently produce an unchanged or partially updated checkpoint when the trained state dict does not match. There are also no automated tests for the slicing/export logic; the stated coverage is only an end-to-end manual run.

File: examples/alpamayo/qad.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+    if args.freeze_vision:
+        frozen = 0
+        for name, param in student_vlm.named_parameters():
+            if "visual" in name:
+                param.requires_grad_(False)
+                frozen += 1
+        print(f"[qad] froze {frozen} vision-tower parameters", flush=True)
+
+    train_ids = clip_slice(args.parquet, args.train_offset, args.limit_train)
+    val_ids = clip_slice(args.parquet, args.val_offset, args.limit_val)
```
[the remark this answers is not in the document]
Comment (rohansjoshi) [reply examples/alpamayo/qad.py:322]: Fixed

File: examples/alpamayo/qad.py (hunk 1 above)
Comment (rohansjoshi) [reply examples/alpamayo/qad.py:240]: Fixed now

File: examples/alpamayo/README.md [hunk 4]
```diff
@@ -68,5 +68,64 @@ candidate format perturbs that loss, then picks the cheapest assignment that
...
+```bash
+# 1. Quantize.
+python quantize.py --ckpt nvidia/Alpamayo-R1-10B --output-dir ./alpamayo-auto --quantize auto
+
+# 2. Distill the quantized VLM against the FP16 one, then export.
+torchrun --standalone --nproc_per_node 8 qad.py \
+    --student_ckpt ./alpamayo-auto \
+    --output_dir ./alpamayo-auto-qad \
+    --parquet ./train_clips.parquet --limit_train 2000 \
+    --max_steps 500 --fsdp2 --grad_ckpt --export
```
Comment (cjluo-nv) [inline examples/alpamayo/README.md:98]: > _Bot comment._

The documented command now fails before training: `--limit_train 2000` selects `[0:2000]`, while the default validation range remains `[12:16]`, and `main()` rejects that overlap. Please include an explicit non-overlapping `--val_offset` here (and ensure the example dataset has enough clips), as requested in the prior review.

File: examples/alpamayo/qad.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+    full = AlpamayoR1.from_pretrained(student_ckpt, dtype=dtype)
+
+    shards = sorted(glob.glob(os.path.join(trained_vlm_dir, "model*.safetensors")))
+    if not shards:
+        raise FileNotFoundError(f"no model*.safetensors found in {trained_vlm_dir}")
+    state_dict = {}
+    for shard in shards:
+        state_dict.update(load_file(shard))
+
+    missing, unexpected = full.vlm.load_state_dict(state_dict, strict=False)
```
Comment (cjluo-nv) [inline examples/alpamayo/qad.py:183]: > _Bot comment._

The strict mismatch handling addresses the export correctness bug, but the previously requested automated coverage is still missing (`tests/examples/alpamayo` only contains `test_quantize.py`). Please add lightweight tests for a successful sharded-state export and for missing/unexpected keys aborting before `save_pretrained`; also cover the train/validation overlap rejection. These paths protect checkpoint integrity and the current README/default regression shows why they need coverage.

**Review by cjluo-nv — COMMENTED** [review]
> _Bot review (gpt-5.6-sol) — DM the bot to share feedback._

Re-review: the rollout documentation is corrected, and export now aborts on missing or unexpected VLM keys. The split fix is only partial: the newly added overlap check makes the README's primary command fail with its default validation range. The previously requested unit coverage for split and export integrity is also still absent. The new file's NVIDIA header matches `LICENSE_HEADER`.

File: tests/examples/alpamayo/test_qad.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        """Test that missing keys in trained state dict trigger error."""
+        missing = {"vlm.model.layers.0.self_attn.q_proj.weight"}
+        unexpected = set()
+
+        # Logic from export_full_model: if missing: raise ValueError
+        has_error = len(missing) > 0
+        assert has_error, "Missing keys should trigger error in export"
+
+    def test_export_unexpected_keys_should_raise(self):
+        """Test that unexpected keys in trained state dict trigger error."""
```
Comment (cjluo-nv) [inline tests/examples/alpamayo/test_qad.py:187]: > _Bot comment._

These tests never import or call `clip_slice`, the split validator, or `export_full_model`; they only assert facts about locally constructed sets and duplicated arithmetic. They would still pass if export stopped checking mismatches or saved despite them, so the previously requested checkpoint-integrity coverage remains unresolved. Please exercise the production code with mocked `AlpamayoR1.from_pretrained`, `glob`/`load_file`, and `save_pretrained`, covering successful multi-shard loading plus missing/unexpected-key failures and asserting that failures occur before save. Likewise, invoke an extracted split-validation helper rather than copying its condition into the test.

File: examples/alpamayo/qad.py [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+        raise ValueError(
+            f"Invalid dtype '{args.dtype}'. Must be a valid torch dtype "
+            f"(e.g., 'float16', 'bfloat16', 'float32')"
+        )
+    export_dir = args.export_dir or f"{args.output_dir.rstrip('/')}-full"
+
+    # Validate train/val split does not overlap
+    train_end = args.train_offset + args.limit_train
+    val_end = args.val_offset + args.limit_val
+    if not (train_end <= args.val_offset or val_end <= args.train_offset):
```
Comment (cjluo-nv) [inline examples/alpamayo/qad.py:288]: > _Bot comment._

Please reject negative offsets/limits before this interval check. For example, `--train_offset 0 --limit_train -1` computes `train_end == -1` and passes as non-overlapping with the default validation range, while `clip_slice(..., 0, -1)` actually selects nearly the entire clip list, including those validation clips. Require offsets to be nonnegative and limits to be nonnegative/positive as appropriate, and cover this case through the production validation helper.

File: tests/examples/alpamayo/test_qad.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
```
Comment (cjluo-nv) [inline tests/examples/alpamayo/test_qad.py:15]: > _Bot comment._

This new source file contains only the two SPDX lines, whereas the repository's canonical `LICENSE_HEADER` (and `test_quantize.py`) includes the full Apache notice. Please use the canonical header. Because this is a GitHub PR touching license notices, the mismatch also requires human/licensing sign-off.

**Review by cjluo-nv — COMMENTED** [review]
> _Bot review (gpt-5.6-sol) — DM the bot to share feedback._

Re-review: the earlier rollout-documentation, README split, and strict export-key issues are now addressed. However, the newly added tests do not execute `qad.py`; they only reimplement its expressions, so the requested critical regression coverage for overlap rejection and export integrity is still absent. The range validation can also be bypassed with negative CLI values, and the new test file's abbreviated SPDX header does not match the repository's canonical `LICENSE_HEADER`.

File: examples/alpamayo/README.md [hunk 9]
```diff
@@ -68,5 +68,65 @@ candidate format perturbs that loss, then picks the cheapest assignment that
...
+feasible). However, rollout sampling and sequence augmentation for generated
+tokens are **not currently implemented** in `qad.py`—only prompt-only targets
+(before `<|cot_start|>`) are supported.
+
+```bash
+# 1. Quantize.
+python quantize.py --ckpt nvidia/Alpamayo-R1-10B --output-dir ./alpamayo-auto --quantize auto
+
+# 2. Distill the quantized VLM against the FP16 one, then export.
+torchrun --standalone --nproc_per_node 8 qad.py \
```
Comment (cjluo-nv) [inline examples/alpamayo/README.md:93]: can we also add some example output of this e.g. printed training loss and validation loss etc? So the user knows what to expect

**Review by cjluo-nv — APPROVED** [review]
